keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+#### controller-manager Startup Configuration for Multiple Service CIDRs
+A new, plural "service-cluster-ip-ranges" option for the [controller-manager startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) is proposed, in addition to retaining the existing, singular "service-cluster-ip-range" option (for backwards compaibility):
+```
+  --service-cluster-ip-range  ipNet         [Singular IP CIDR,  Default: 10.0.0.0/24]
+  --service-cluster-ip-ranges stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --service-cluster-ip-range argument will become deprecated.
+
+#### 'kubectl get service' Command Display for Multiple Service IPs
+The 'kubectl get service ...' command display will need to be modified to return multiple service IPs, e.g.:
+```
+       kube-master# kubectl get service
+       NAME         TYPE        CLUSTER-IP               EXTERNAL-IP   PORT(S)   AGE
+       kubernetes   ClusterIP   fd00:1234::1,10.96.0.1   <none>        443/TCP   20m
+       kube-master#
+```
+
+### Endpoints
+
+For the first release of dual-stack support, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. The addresses that are included in the "Subsets" field of an "Endpoints" object will continue to be a list of separate addresses, rather than changing this to a list of dual-stack N-tuples of addresses.
+
+Behaviorly, this means that each IPv4/IPv6 address that is assigned to a backend pod will be treated as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+
+#### 'kubectl get endpoints' Command Display for Dual-Stack Backend Pods
+There should be no changes required for the 'kubectl get endpoints ...' command display. The command output should include both IPv4 and IPv6 addresses as individual endpoints, e.g.:
+```
+       kube-master# kubectl get endpoints
+       NAME         ENDPOINTS                         AGE
+       kubernetes   [fd00:90::2]:6443,10.0.0.2:6643   20m
+       kube-master#
+```
+
+### kube-proxy Operation
+
+In order to support dual-stack Kubernetes services, kube-proxy will be modified to support parallel support for iptables and ip6tables mappings. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to iptables and ip6tables in parallel.
+
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+New [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) arguments will be added to allow a user to specify separate IPv4 and IPv6 bind addresses for the kube-proxy server:
```

Comment:
Mentioned above, would it suffice to just support health checks for one IP mode (IPv6, if available, else IPv4)?

For metrics, wouldn't only one IP mode be needed to access the metrics? Not sure why one would want to access metrics from IPv4 and IPv6. May be able to leave that, as is.